include/torch-mlir/Dialect/Torch/IR/TorchOps.h

```diff
@@ -294,6 +294,35 @@ bool isListPotentiallyMutated(Value list);
 /// the list.
 bool potentiallyMutatesListOperands(Operation *op);
 
+/// Returns the value from an `IntegerAttr` as an `int64_t`.
+///
+/// @param intAttr the `IntegerAttr` from which to extract the value
+/// @return the value as an `int64_t`
+///
+/// Regardless of the signed-ness of the attribute, this function returns
+/// the value as a signed integer, which implies that if the attribute has
+/// a 64-bit unsigned value, it will be converted to an int64_t in the manner
+/// that uint64_t is cast to int64_t in C++.
+int64_t getIntAttrAsSigned(IntegerAttr intAttr);
+
+/// Returns the value from an `IntegerAttr` as an integral index.
```

Comment:
Does torch always allow indices in `[-ndim, ndim)`, does it maybe sometimes allow any value and just modulo it ndim? I don't know, but if that's the case this util might trip a developer up. 

In my PR to this (coming soon) I've slightly simplified these utils and made them private to the .cpp -- use if you like.

Reply:
No, modulo is not the behavior.  It gives an "INDICES element is out of DATA bounds" error.  Of course, the pytorch docs don't specify these things, not even about negative indices.

Thanks for the rewrite.  I can take the simplification of the casting logic.  However, you've collapsed my two levels of abstraction into one for some reason I don't understand.  The lower level, getIntAttrAsSigned, was intended for applications where you know the usage of an IntegerAttr is signed but the MLIR could be written as any signed-ness.  The upper level, getIntAttrAsIndex, is for those cases when you know that your integer is specifically a tensor index.

You've also pushed these utilities down, when I think the opposite should be the case: it should be as public as possible, even in IntegerAttr itself.  However, I think taking on a change to the MLIR API itself might cause this PR to drag on too long, so I'm hoping at least to encourage this kind of abstraction among users of TorchOps.h.

Reply:
> No, modulo is not the behavior. It gives an "INDICES element is out of DATA bounds" error. Of course, the pytorch docs don't specify these things, not even about negative indices.

 I was trying to ask if ALL torch ops in torch have this behavior, and if not then this API is not useful for the cases where it might by modulo logic. 
 
 > However, you've collapsed my two levels of abstraction into one for some reason I don't understand.
 
Personal taste I suppose, maybe I "inline heuristic" is more aggressive.

Reply:
> I was trying to ask if ALL torch ops in torch have this behavior, and if not then this API is not useful for the cases where it might by modulo logic.
Oh, I see.  Yeah, pytorch doesn't document at this level, which means someone would have to test all of them, and it could change at any time.  Not an ideal way to design an API.

However, I don't think the value of getIntAttrAsIndex is diminished from not having consistent API behavior, as the function itself has well-defined behavior.  That is, you use the function only when it matches what you want.  If it is discovered that there are multiple behaviors for indexes, it would be better to enhance getIntAttrAsIndex.  For example, if you have an op with modulo behavior, then perhaps a bool usesModulo arg should be added.  

I think the current technique of ad hoc implementations all over the code, some matching the behavior of pytorch and some not, is far worse.  I don't see these implementations even considering corner cases like modulo or negative indexes.  I feel that lowering the effort through the addition of the right abstractions is the best way to deal with that problem.  I don't know that I've come up with the right abstractions, but at least I want to start that conversation.

> Personal taste I suppose, maybe I "inline heuristic" is more aggressive.
Inlining the implemenation would be better; I'll do that.  However, I'm hoping to turn your attention away from the style of implementation and toward my point that there are two levels of abstractions that would be useful.  Perhaps you disagree that the lower level is useful, and in fact, I don't personally have a second use case for it yet.

I'll have to admit that my efforts to encourage others to introduce levels of abstraction for higher quality and better functional coverage have largely failed over the decades.  I introduce the abstractions and write code to demonstrate how using those abstractions raises the level at which one implements, but everyone still continues to code at the level of the underlying API.  It doesn't stop me from trying, though!  I believe my difficulty is in getting others to recognize the opportunity for abstraction in their own code; it takes a lot of mentoring, which I could really only do with my own staff.

Reply:
done according to offline agreement